### ISSUE TITLE
Xerox 2.0 Bug in this script

### ISSUE THREAD

**Issue by tobltobs** [jasonlfunk/ocr-text-extraction#2]
The script makes some funny things with for e.g. an "8":

![original](https://f.cloud.github.com/assets/1296747/968854/93909a6a-05b0-11e3-81c6-aba03e04a7c5.jpg)

![out](https://f.cloud.github.com/assets/1296747/968851/8a2e9e4a-05b0-11e3-943f-2863f0af41ab.png)

**Comment by jasonlfunk** [comment]
Yeah. I know the script doesn't work very well. I'm not actively supporting it.

**Comment by RusEu** [comment]
same problem here :(

**Comment by RusEu** [comment]
Also with B letter ..

**Comment by RusEu** [comment]
On commenting this:
    # if count_children(index, h_, contour) > 2:
    #     if DEBUG:
    #         print "\t skipping, is a container of letters"
    #     return True
I get the B and the 8 , but very dark and the OCR doesn't recognize it well. Any idea how to solve this?

**Comment by varunkumar2310** [comment]
max edge-boxes that can occur inside a text edge-box = 2
but each edge-box has 2 contours : outer contour + inner contour
take case of A : total contours = outer contour of A + inner contour of A + outer contour of delta + inner contour of delta = 4 contours
And an edge-box can had max of 2 edge-boxes completely enclosed in it.
(max enclosing edge-boxes for English language = 2)
Which means max contours that can be enclosed inside a text ( inside outer contour of text )
= inner contour of text + 2 \* max edge-boxes = 1 + 2 \* 2 = 5
change the max contours depending on language if necessary

Therefore in include_box:
modify the lines to :
if is_child(index, h_) and count_children(get_parent(index, h_), h_, contour) <= 5:
if count_children(index, h_, contour) > 5:

I have coded in OpenCV 3.1, C++ , and the following change works for me.
Correct me if my logic is wrong.

Images:
original
![testing](https://cloud.githubusercontent.com/assets/22290813/19869053/aad7cc26-9fd0-11e6-9b7e-3bb5a4b02027.png)
processed_with_5
![processed_with_5](https://cloud.githubusercontent.com/assets/22290813/19869029/9c1363e4-9fd0-11e6-8a02-7632c14c3497.png)
rejected_with_5
![rejected_with_5](https://cloud.githubusercontent.com/assets/22290813/19869034/9f4f1170-9fd0-11e6-9bff-34067c9e0c36.png)
processed_with_2
![processed_with_2](https://cloud.githubusercontent.com/assets/22290813/19869040/a2e03986-9fd0-11e6-9523-c3ca57b55357.png)
rejected_with_2
![rejected_with_2](https://cloud.githubusercontent.com/assets/22290813/19869045/a6c41a68-9fd0-11e6-9d23-9218347f4084.png)

**Comment by guddulrk** [comment]
I am getting following error:

 inf = file(scratch_text_name_root + '.txt')
NameError: name 'file' is not defined

Any help please??

**Comment by tsjason** [comment]
@guddulrk No idea. It looks like you might be missing some libraries or something. But I haven't used this in ages so I'm not sure.

**Comment by itcthienkhiem** [comment]
same error.

**Comment by itcthienkhiem** [comment]
![capture](https://user-images.githubusercontent.com/2140355/29093780-3d9fa09e-7cb5-11e7-8c25-149d34ee63d7.JPG)
B still problem ?
How can fix it ?

**Comment by NightFury13** [comment]
> max edge-boxes that can occur inside a text edge-box = 2
> but each edge-box has 2 contours : outer contour + inner contour
> take case of A : total contours = outer contour of A + inner contour of A + outer contour of delta + inner contour of delta = 4 contours
> And an edge-box can had max of 2 edge-boxes completely enclosed in it.
> (max enclosing edge-boxes for English language = 2)
> Which means max contours that can be enclosed inside a text ( inside outer contour of text )
> = inner contour of text + 2 * max edge-boxes = 1 + 2 * 2 = 5
> change the max contours depending on language if necessary
> 
> Therefore in include_box:
> modify the lines to :
> if is_child(index, h_) and count_children(get_parent(index, h_), h_, contour) <= 5:
> if count_children(index, h_, contour) > 5:
> 
> I have coded in OpenCV 3.1, C++ , and the following change works for me.
> Correct me if my logic is wrong.
> 
> Images:
> original
> ![testing](https://cloud.githubusercontent.com/assets/22290813/19869053/aad7cc26-9fd0-11e6-9b7e-3bb5a4b02027.png)
> processed_with_5
> ![processed_with_5](https://cloud.githubusercontent.com/assets/22290813/19869029/9c1363e4-9fd0-11e6-8a02-7632c14c3497.png)
> rejected_with_5
> ![rejected_with_5](https://cloud.githubusercontent.com/assets/22290813/19869034/9f4f1170-9fd0-11e6-9bff-34067c9e0c36.png)
> processed_with_2
> ![processed_with_2](https://cloud.githubusercontent.com/assets/22290813/19869040/a2e03986-9fd0-11e6-9523-c3ca57b55357.png)
> rejected_with_2
> ![rejected_with_2](https://cloud.githubusercontent.com/assets/22290813/19869045/a6c41a68-9fd0-11e6-9d23-9218347f4084.png)

Pappu! Thanks dude! :D @varunkumar2310

**Comment by sstefanov** [comment]
Because of used Canny each contour is recognized twice.
This is the reason for not proper handling B and 8 chars. Children are 4, not 2.

To fix this just replace 2 with 4 in:
if is_child(index, h_) and count_children(get_parent(index, h_), h_, contour) <= 2:
must be
if is_child(index, h_) and count_children(get_parent(index, h_), h_, contour) <= 4:
and
if count_children(index, h_, contour) > 2:
must be
if count_children(index, h_, contour) > 4:

See my pull request.
